components/homme/src/share/cxx/utilities/MathUtils.hpp

```diff
@@ -36,6 +36,54 @@ KOKKOS_INLINE_FUNCTION constexpr FPType max(const FPType val, FPPack... pack) {
   return val > max(pack...) ? val : max(pack...);
 }
 
+template <typename FPType>
+KOKKOS_INLINE_FUNCTION constexpr FPType square(const FPType& x) {
+  return x*x;
+}
+
+// Set the least significant n bits of the fraction to 0. Use this for
+// BFB-testing between GPU and host. Table 8 of Appendix E of the Cuda
+// programming guide gives the accuracy of various functions in base
+// 10. However, we also have to account for carries in the calculations. Thus,
+// even if an answer is accurate to 2 base-10 UL or ~7 bits, carries can make
+// more bits differ. Thus, zeroing (or any sort of rounding) is not assured to
+// work. However, you can reduce the probability of a diff to a small number by
+// choosing n to be large. Full double precision range is 53 bits; choose
+// something like n = 20 to get a carry-based diff every 1 in a million calls,
+// still leaving 33 bits.
+KOKKOS_INLINE_FUNCTION
+double zeroulpn (const double& a, const int n) {
+  // Use frexp, lexp, and bit shifts on a long long int representation to avoid
+  // manipulating bits of the FP number.
+  int e;
+  auto xd = std::frexp(a, &e);
+  const int sign = xd >= 0 ? 1 : -1;
+  xd = std::abs(xd);
+  const auto f = static_cast<double>(1LL << 54);
+  long long int x = f*xd;
+  x >>= n;
+  x <<= n;
+  xd = x/f;
+  return std::ldexp(sign*xd, e);
+}
+
+template <typename Pack>
+KOKKOS_INLINE_FUNCTION
+void zero_ulp_n (Pack& a, const int n) {
+  for (int s = 0; s < VECTOR_SIZE; ++s) a[s] = zeroulpn(a[s], n);
+}
+
+template <typename Pack>
+KOKKOS_INLINE_FUNCTION
+Pack bfb_pow (const Pack& a, const Real& e) {
+  Pack b;
+  for (int s = 0; s < VECTOR_SIZE; ++s)
+    b[s] = std::pow(a[s], e);
+  if (OnGpu<ExecSpace>::value)
+    zero_ulp_n(b, 20);
```

Comment:
The Cuda programming guide states that the max ulp error for pow is 2. Is there a reason why we should use 20 instead of 2? It may mask some other upstream non bfb errors.

Reply:
Keep in mind this function is meant to apply to the single pow call. So all it's doing is making pow less accurate. That won't mask other errors, since all other operations are done at full precision and are not modified.

As for 20 base-2 digts vs 2 base-10 digits ~ 7 base-2 digits, the issue is carrying. I quote the doc for zeroulpn:
```C++
// Set the least significant n bits of the fraction to 0. Use this for
// BFB-testing between GPU and host. Table 8 of Appendix E of the Cuda
// programming guide gives the accuracy of various functions in base
// 10. However, we also have to account for carries in the calculations. Thus,
// even if an answer is accurate to 2 base-10 UL or ~7 bits, carries can make
// more bits differ. Thus, zeroing (or any sort of rounding) is not assured to
// work. However, you can reduce the probability of a diff to a small number by
// choosing n to be large. Full double precision range is 53 bits; choose
// something like n = 20 to get a carry-based diff every 1 in a million calls,
// still leaving 33 bits.
```

Reply:
Ah, I see. I thought CUDA's API was in bits, since gamma has a max error of 11ulps, which seemed "large" if given in decimal digits.

When I said "mask other ops", I meant that if the base of the pow had 10 wrong bits (due to other non bfb stuff), zeroing out 20 bits would not let you see that diff.

Anyhow, my understanding of CUDA programming guide was wrong, so nvm.

Reply:
No, I'm wrong: I think you're right about the base-2 interpretation. They write, e.g., 2 [ULP] = 2*ulp, with ulp something like `ldexp(2,-53)`. So a base-2 interpretation is correct. I'll modify the zeroulpn doc in my next PR to reflect that.

As for masking other ops through a reduced-precision pow (for example), that's a really interesting point. I have an idea to fix that. Let x, n be the arguments to zeroulpn, and let y be the result. zeroulpn will zero the last n bits of the fraction of y. Then replace those with the last n bits of x. That's an n-bit perturbation that will require x to be BFB in those least significant digits.

Reply:
Nice idea. No rush on that, we can put it in with dirk-phase-2.